proof.go

```diff
@@ -465,6 +465,202 @@ func (proof Proof) VerifyInclusion(h hash.Hash, nid namespace.ID, leavesWithoutN
+	rootHash, err := computeRoot(0, proofRangeSubtreeEstimate)
+	if err != nil {
+		return false, fmt.Errorf("failed to compute root [%d, %d): %w", 0, proofRangeSubtreeEstimate, err)
+	}
+	for i := 0; i < len(proof.Nodes()); i++ {
+		rootHash, err = nth.HashNode(rootHash, proof.Nodes()[i])
+		if err != nil {
+			return false, fmt.Errorf("failed to hash node: %w", err)
+		}
+	}
+
+	return bytes.Equal(rootHash, root), nil
+}
+
+// ToLeafRanges returns the leaf ranges corresponding to the provided subtree roots.
+// The proof range defined by proofStart and proofEnd is end exclusive.
+// It uses the subtree root width to calculate the maximum number of leaves a subtree root can
+// commit to.
+// The subtree root width is defined as per ADR-013:
+// https://github.com/celestiaorg/celestia-app/blob/main/docs/architecture/adr-013-non-interactive-default-rules-for-zero-padding.md
+// This method assumes:
+// - The subtree roots are created according to the ADR-013 non-interactive defaults rules
+// - The tree's number of leaves is a power of two
+// The algorithm is as follows:
+// - Let `d` be `y - x` (the range of the proof).
+// - `i` is the index of the next subtree root.
+// - While `d != 0`:
+//   - Let `z` be the largest power of 2 that fits in `d`; here we are finding the range for the next subtree root.
+//   - The range for the next subtree root is `[x, x + z)`, i.e., `S_i` is the subtree root of leaves at indices `[x, x + z)`.
+//   - `d = d - z` (move past the first subtree root and its range).
+//   - `i = i + 1`.
+//   - Go back to the loop condition.
+//
+// Note: This method is Celestia specific.
+func ToLeafRanges(proofStart, proofEnd, subtreeWidth int) ([]LeafRange, error) {
+	if proofStart < 0 {
+		return nil, fmt.Errorf("proof start %d shouldn't be strictly negative", proofStart)
+	}
+	if proofEnd <= proofStart {
+		return nil, fmt.Errorf("proof end %d should be stricly bigger than proof start %d", proofEnd, proofStart)
+	}
+	if subtreeWidth <= 0 {
+		return nil, fmt.Errorf("subtree root width cannot be negative %d", subtreeWidth)
+	}
+	currentStart := proofStart
+	currentLeafRange := proofEnd - proofStart
+	var ranges []LeafRange
+	maximumLeafRange := subtreeWidth
+	for currentLeafRange != 0 {
+		nextRange, err := nextLeafRange(currentStart, proofEnd, maximumLeafRange)
+		if err != nil {
+			return nil, err
+		}
+		ranges = append(ranges, nextRange)
+		currentStart = nextRange.End
+		currentLeafRange = currentLeafRange - nextRange.End + nextRange.Start
+	}
+	return ranges, nil
+}
+
+// nextLeafRange takes a proof start, proof end, and the maximum range a subtree
+// root can cover, and returns the corresponding subtree root range.
+// Check ToLeafRanges() for more information on the algorithm used.
+// The subtreeWidth is calculated using SubTreeWidth() method
+// in celestiaorg/go-square/inclusion package.
+// Note: This method is Celestia specific.
+func nextLeafRange(currentStart, currentEnd, subtreeWidth int) (LeafRange, error) {
```

Comment:
It is good to indicate in the go doc that the `subtreeWidth` is expected to be a power of two. Additionally, the `LeafRange` values will all be powers of two.

Reply:
what do you mean by the leaf range values? you mean start and end? or the range size?

Reply:
By `LeafRange` I meant range size.

Reply:
[3ad98af](https://github.com/celestiaorg/nmt/pull/260/commits/3ad98af20ba715d54aab08441e0129406161f49b)